File: data/convert_to_backend_seeds.py
#!/usr/bin/env python3
"""
Convert data/seed_data.json + data/invoices.json (C00X IDs)
→ backend/seeds/seed_data.json + backend/seeds/invoices.json (UUID format)

UUID mapping: C001 → f47ac10b-0001-4000-8000-000000000000
              C010 → f47ac10b-000a-4000-8000-000000000000
              C050 → f47ac10b-0032-4000-8000-000000000000
"""
import json
import os
import re
from pathlib import Path

ROOT = Path(__file__).parent.parent
DATA_DIR = Path(__file__).parent
SEEDS_DIR = ROOT / "backend" / "seeds"


def cid_to_uuid(cid: str) -> str:
    """Convert 'C001' → 'f47ac10b-0001-4000-8000-000000000000'"""
    n = int(cid[1:])
    return f"f47ac10b-{n:04x}-4000-8000-000000000000"


def convert_companies():
    with open(DATA_DIR / "seed_data.json") as f:
        companies = json.load(f)

    out = {"companies": []}
    for c in companies:
        out["companies"].append({
            "id": cid_to_uuid(c["id"]),
            "name": c["name"],
            "sector": c["sector"],
            "attributes": {
                "subtype": c.get("subtype"),
                "district": c.get("district"),
                "gls_member": c.get("gls_member", False),
                "founded": c.get("founded"),
                "size": c.get("size"),
            },
        })

    target = SEEDS_DIR / "seed_data.json"
    with open(target, "w") as f:
        json.dump(out, f, indent=2, ensure_ascii=False)
    print(f"✓ {target} written ({len(out['companies'])} companies)")
    return out["companies"]


def convert_invoices():
    with open(DATA_DIR / "invoices.json") as f:
        invoices = json.load(f)

    # Sector hint lookup for descriptions
    with open(DATA_DIR / "seed_data.json") as f:
        companies_raw = json.load(f)
    sector_by_cid = {c["id"]: c["sector"] for c in companies_raw}

    SECTOR_DESC = {
        "port_logistics": [
            "Frachtabwicklung", "Zollabfertigung Containerimport",
            "Lagereingang Sammelgut", "Warenausgabe", "Container-Spedition",
            "Frachtkosten Rücktransport", "Lagernutzung", "Umschlaggebühr",
            "Hafengebühr THC", "Speicherung Frachtgut",
        ],
        "food_beverage": [
            "Warenlieferung Lebensmittel", "Rohstofflieferung",
            "Gastronomieausrüstung", "Braudienstleistungen", "Catering-Service",
            "Transportkosten Lebensmittel", "Qualitätsprüfung Waren",
        ],
        "renewable_energy": [
            "Stromerzeugung Windkraft", "Photovoltaik-Wartung",
            "Netzanbindungsgebühr", "Speichermiete Batteriespeicher",
            "Regelenergie-Dienstleistung", "Ertragsgutachten Solaranlage",
        ],
    }
    DEFAULT_DESCS = ["Dienstleistung B2B", "Warenlieferung", "Beratungsleistung"]

    def get_desc(from_cid):
        sector = sector_by_cid.get(from_cid, "")
        descs = SECTOR_DESC.get(sector, DEFAULT_DESCS)
        import hashlib
        h = int(hashlib.md5(from_cid.encode()).hexdigest(), 16)
        return descs[h % len(descs)]

    out_invoices = []
    for i, inv in enumerate(invoices, 1):
        from_cid = inv["from_company_id"]
        to_cid = inv["to_company_id"]

        # Convert line items: unit_price_cents → amount_cents
        line_items = []
        for li in inv.get("line_items", []):
            qty = li.get("quantity", 1)
            unit = li.get("unit_price_cents", 0)
            li_total = unit * qty
            line_items.append({
                "description": li["description"],
                "amount_cents": li_total,
                "quantity": qty,
            })

        out_invoices.append({
            "id": f"INV-{i:04d}",
            "from_company_id": cid_to_uuid(from_cid),
            "to_company_id": cid_to_uuid(to_cid),
            "amount_cents": inv["amount_cents"],
            "description": get_desc(from_cid),
            "due_date": inv.get("due_date", "2026-06-30"),
            "status": "pending",
            "line_items": line_items,
        })

    target = SEEDS_DIR / "invoices.json"
    with open(target, "w") as f:
        json.dump({"invoices": out_invoices}, f, indent=2, ensure_ascii=False)
    print(f"✓ {target} written ({len(out_invoices)} invoices)")

    # Quick netting analysis
    gross = sum(inv["amount_cents"] for inv in out_invoices)
    bilateral = {}
    for inv in out_invoices:
        a, b = inv["from_company_id"], inv["to_company_id"]
        key = tuple(sorted([a, b]))
        if key not in bilateral:
            bilateral[key] = {a: 0, b: 0}
        bilateral[key][a] = bilateral[key].get(a, 0) + inv["amount_cents"]
    nettable = sum(min(v.values()) for v in bilateral.values() if len(v) == 2)
    pct = nettable / gross * 100 if gross else 0
    print(f"  Gross: €{gross/100:,.0f}  Nettable: €{nettable/100:,.0f}  Savings: {pct:.1f}%")


if __name__ == "__main__":
    convert_companies()
    convert_invoices()

File: data/convert_to_backend_seeds.js
#!/usr/bin/env node
/**
 * Convert data/seed_data.json + data/invoices.json (C00X IDs)
 * → backend/seeds/seed_data.json + backend/seeds/invoices.json (UUID format)
 *
 * UUID mapping: C001 → f47ac10b-0001-4000-8000-000000000000
 *               C010 → f47ac10b-000a-4000-8000-000000000000
 *               C050 → f47ac10b-0032-4000-8000-000000000000
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

const DATA_DIR = __dirname;
const SEEDS_DIR = path.join(__dirname, '..', 'backend', 'seeds');

function cidToUuid(cid) {
  const n = parseInt(cid.slice(1), 10);
  const hex = n.toString(16).padStart(4, '0');
  return `f47ac10b-${hex}-4000-8000-000000000000`;
}

// ── Convert companies ────────────────────────────────────────────────────────
const companies = JSON.parse(fs.readFileSync(path.join(DATA_DIR, 'seed_data.json'), 'utf8'));

const seedCompanies = {
  companies: companies.map(c => ({
    id: cidToUuid(c.id),
    name: c.name,
    sector: c.sector,
    attributes: {
      subtype: c.subtype || null,
      district: c.district || null,
      gls_member: c.gls_member || false,
      founded: c.founded || null,
      size: c.size || null,
    },
  })),
};

fs.writeFileSync(
  path.join(SEEDS_DIR, 'seed_data.json'),
  JSON.stringify(seedCompanies, null, 2)
);
console.log(`✓ backend/seeds/seed_data.json written (${seedCompanies.companies.length} companies)`);

// ── Convert invoices ─────────────────────────────────────────────────────────
const invoicesRaw = JSON.parse(fs.readFileSync(path.join(DATA_DIR, 'invoices.json'), 'utf8'));

const sectorByCid = {};
companies.forEach(c => { sectorByCid[c.id] = c.sector; });

const SECTOR_DESC = {
  port_logistics: [
    'Frachtabwicklung', 'Zollabfertigung Containerimport',
    'Lagereingang Sammelgut', 'Warenausgabe Hanseatic',
    'Container-Spedition FCL', 'Frachtkosten Rücktransport',
    'Lagernutzung Zollware', 'Umschlaggebühr', 'Hafengebühr THC',
    'Speicherung Frachtgut Q1',
  ],
  food_beverage: [
    'Warenlieferung Lebensmittel', 'Rohstofflieferung Brauerei',
    'Gastronomieausrüstung', 'Braudienstleistungen März',
    'Catering-Service', 'Transportkosten Lebensmittel',
    'Qualitätsprüfung Waren',
  ],
  renewable_energy: [
    'Stromerzeugung Windkraft', 'Photovoltaik-Wartung',
    'Netzanbindungsgebühr April', 'Speichermiete Batteriespeicher',
    'Regelenergie-Dienstleistung', 'Ertragsgutachten Solaranlage',
  ],
};

function getDesc(fromCid) {
  const sector = sectorByCid[fromCid] || 'cross_cluster';
  const descs = SECTOR_DESC[sector] || ['Dienstleistung B2B', 'Warenlieferung', 'Beratungsleistung'];
  const h = parseInt(crypto.createHash('md5').update(fromCid).digest('hex').slice(0, 8), 16);
  return descs[h % descs.length];
}

const invoices = {
  invoices: invoicesRaw.map((inv, i) => {
    const lineItems = (inv.line_items || []).map(li => ({
      description: li.description,
      amount_cents: (li.unit_price_cents || 0) * (li.quantity || 1),
      quantity: li.quantity || 1,
    }));

    return {
      id: `INV-${String(i + 1).padStart(4, '0')}`,
      from_company_id: cidToUuid(inv.from_company_id),
      to_company_id: cidToUuid(inv.to_company_id),
      amount_cents: inv.amount_cents,
      description: getDesc(inv.from_company_id),
      due_date: inv.due_date || '2026-06-30',
      status: 'pending',
      line_items: lineItems,
    };
  }),
};

fs.writeFileSync(
  path.join(SEEDS_DIR, 'invoices.json'),
  JSON.stringify(invoices, null, 2)
);
console.log(`✓ backend/seeds/invoices.json written (${invoices.invoices.length} invoices)`);

// ── Netting sanity check ────────────────────────────────────────────────────
const gross = invoices.invoices.reduce((s, inv) => s + inv.amount_cents, 0);
const bilateral = {};
invoices.invoices.forEach(inv => {
  const key = [inv.from_company_id, inv.to_company_id].sort().join('|');
  if (!bilateral[key]) bilateral[key] = {};
  bilateral[key][inv.from_company_id] = (bilateral[key][inv.from_company_id] || 0) + inv.amount_cents;
});
const nettable = Object.values(bilateral)
  .filter(v => Object.keys(v).length === 2)
  .reduce((s, v) => s + Math.min(...Object.values(v)), 0);
const pct = (nettable / gross * 100).toFixed(1);

console.log(`\nNetting sanity:`);
console.log(`  Gross:    €${(gross / 100).toLocaleString('de-DE')}`);
console.log(`  Nettable: €${(nettable / 100).toLocaleString('de-DE')}`);
console.log(`  Savings:  ${pct}%`);
console.log(pct >= 25 ? '  ✅ Meets ≥25% Fleischman benchmark' : '  ⚠ Below 25% benchmark');
